File: iOS/DittoPOS/Data/Repositories/OrdersRepository.swift
//
//  OrdersRepository.swift
//  DittoPOS
//
//  Copyright © 2026 DittoLive Incorporated. All rights reserved.
//

import Combine
import DittoSwift
import Foundation

/// Owns the `pos_orders` collection: registers the per-location subscription,
/// observes the synced documents, runs DQL mutations, and gates startup
/// eviction. Mirrors Android's `OrdersRepository`.
@MainActor final class OrdersRepository: ObservableObject {

    @Published private(set) var locationOrders: [Order] = []

    private let dittoManager: DittoManager
    private var ditto: Ditto { dittoManager.ditto }
    private var store: DittoStore { ditto.store }
    private var sync: DittoSync { ditto.sync }

    private var subscription: DittoSyncSubscription?
    private var observer = AnyCancellable({})
    private var locationSubscription = AnyCancellable({})

    init(dittoManager: DittoManager = .shared, locationsRepository: LocationsRepository) {
        self.dittoManager = dittoManager
        // Pull the active location from LocationsRepository and re-register
        // the subscription on every change. @Published replays the current
        // value on subscribe, so cold-start activation comes through here.
        locationSubscription = locationsRepository.$currentLocationId
            .compactMap { $0 }
            .removeDuplicates()
            .sink { [weak self] locationId in
                self?.setActiveLocation(locationId)
            }
    }

    // MARK: Subscription

    private func setActiveLocation(_ locationId: String) {
        subscription?.cancel()
        let args: [String: Any?] = ["locationId": locationId, "TTL": DateFormatter.startOfTodayString]

        // Subscription = the sync set: this location's orders since TTL. A failed
        // registration must NOT skip the observer below — the observer is a local
        // read over the store, independent of the sync set, so the UI still shows
        // documents already present locally (matches Android's separation).
        do {
            subscription = try sync.registerSubscription(
                query: """
                    SELECT * FROM \(Order.collectionName)
                    WHERE _id.locationId = :locationId
                        AND createdAt > :TTL
                    """,
                arguments: args
            )
        } catch {
            reportSubscriptionFailure("subscribe orders", error)
        }

        // Observer = the local read that drives the UI. Its own query string,
        // independent of the subscription's success.
        observer = store.observePublisher(
            query: """
                SELECT * FROM \(Order.collectionName)
                WHERE _id.locationId = :locationId
                    AND createdAt > :TTL
                """,
            arguments: args,
            mapTo: Order.self
        )
        .replaceError(with: [])
        .assign(to: \.locationOrders, on: self)
    }

    // MARK: Single-order observation (used by KDS tile views)

    func orderPublisher(_ order: Order) -> AnyPublisher<Order, Never> {
        store.observePublisher(
            query: """
                SELECT * FROM \(Order.collectionName)
                WHERE _id.id = :id AND _id.locationId = :locationId
                """,
            arguments: ["id": order.documentId.id, "locationId": order.documentId.locationId],
            mapTo: Order.self
        )
        .compactMap(\.first)
        .catch { _ in Empty<Order, Never>() }
        .eraseToAnyPublisher()
    }

    // MARK: Mutations

    func add(order: Order) {
        upsert(order: order)
    }

    func add(item: CartLineItem, lineItemId: String, to order: Order) {
        upsert(order: order.addingCartLineItem(item, lineItemId: lineItemId))
    }

    func updateStatus(of order: Order, with status: OrderStatus) {
        upsert(order: order.appendingStatus(status))
    }

    func addPayment(_ payment: Payment, paymentId: String, to order: Order) {
        upsert(order: order.addingPayment(payment, paymentId: paymentId))
    }

    func clearCart(of order: Order) {
        guard !order.cart.isEmpty else { return }
        // UNSET target paths can't be parameterized in DQL, so the cart keys
        // (app-generated line-item UUIDs) are interpolated; all values use :named args.
        let unsetList = order.cart.keys.map { "cart.\"\($0)\"" }.joined(separator: ", ")
        execute(
            """
            UPDATE \(Order.collectionName)
            UNSET \(unsetList)
            WHERE _id.id = :id AND _id.locationId = :locationId
            """,
            args: ["id": order.documentId.id, "locationId": order.documentId.locationId]
        )
    }

    func reset(order: Order) {
        let createdAtNow = DittoWireDate.string(from: Date())
        let args: [String: Any?] = [
            "id": order.documentId.id,
            "locationId": order.documentId.locationId,
            "createdAt": createdAtNow
        ]
        let setClause = "SET createdAt = :createdAt"
        let whereClause = "WHERE _id.id = :id AND _id.locationId = :locationId"

        if order.cart.isEmpty {
            execute("UPDATE \(Order.collectionName) \(setClause) \(whereClause)", args: args)
        } else {
            let unsetList = order.cart.keys.map { "cart.\"\($0)\"" }.joined(separator: ", ")
            execute(
                "UPDATE \(Order.collectionName) \(setClause) UNSET \(unsetList) \(whereClause)",
                args: args
            )
        }
    }

    // MARK: Eviction
    //
    // Storage cleanup on app launch, gated to at most once per 24h. Observer
    // queries filter by location/TTL, so this is purely about preventing the
    // local store from accumulating expired orders.

    func runEvictionIfDue() async {
        let now = Date().timeIntervalSince1970
        let last = UserDefaults.standard.double(forKey: Eviction.lastRunKey)
        guard now - last >= Eviction.interval else { return }

        let ttl = DateFormatter.startOfTodayString
        do {
            _ = try await store.execute(
                query: "EVICT FROM \(Order.collectionName) WHERE createdAt <= :TTL",
                arguments: ["TTL": ttl]
            )
            UserDefaults.standard.set(now, forKey: Eviction.lastRunKey)
            print("Eviction: evicted orders with createdAt <= \(ttl)")
        } catch {
            print("Eviction: ERROR \(error.localizedDescription)")
        }
    }

    private enum Eviction {
        static let lastRunKey = "v2.lastEvictionAt"
        static let interval: TimeInterval = 60 * 60 * 24
    }

    // MARK: Helpers

    private func upsert(order: Order) {
        guard let json = try? order.dittoJSONString() else { return }
        execute(
            """
            INSERT INTO \(Order.collectionName)
            DOCUMENTS (deserialize_json(:json))
            ON ID CONFLICT DO UPDATE_LOCAL_DIFF
            """,
            args: ["json": json]
        )
    }

    private func execute(_ query: String, args: [String: Any?] = [:], function: String = #function) {
        Task {
            do {
                _ = try await store.execute(query: query, arguments: args)
            } catch {
                assertionFailure("DQL \(function) failed: \(error.localizedDescription)\n\(query)")
            }
        }
    }
}

File: Android/app/src/main/java/live/ditto/pos/core/data/repository/OrdersRepository.kt
package live.ditto.pos.core.data.repository

import android.content.Context
import android.util.Log
import com.ditto.kotlin.Ditto
import com.ditto.kotlin.DittoSyncSubscription
import dagger.hilt.android.qualifiers.ApplicationContext
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.distinctUntilChanged
import kotlinx.coroutines.flow.filter
import kotlinx.coroutines.flow.launchIn
import kotlinx.coroutines.flow.onEach
import kotlinx.datetime.Clock
import kotlinx.datetime.TimeZone
import kotlinx.datetime.atStartOfDayIn
import kotlinx.datetime.toLocalDateTime
import live.ditto.ditto_wrapper.DittoManager
import live.ditto.pos.core.data.dittoJsonString
import live.ditto.pos.core.data.observeAsFlow
import live.ditto.pos.core.data.orders.Order
import live.ditto.pos.core.data.toDittoIsoString
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class OrdersRepository @Inject constructor(
    @ApplicationContext private val context: Context,
    private val dittoManager: DittoManager,
    private val locationsRepository: LocationsRepository
) {
    private val ditto: Ditto get() = dittoManager.requireDitto()
    private var subscription: DittoSyncSubscription? = null
    private val repoScope = CoroutineScope(SupervisorJob() + Dispatchers.IO)

    init {
        // Pull the active location from LocationsRepository and re-register
        // the subscription on every change. Mirrors iOS.
        locationsRepository.locationIdFlow()
            .filter { it.isNotEmpty() }
            .distinctUntilChanged()
            .onEach { locationId -> setActiveLocation(locationId) }
            .launchIn(repoScope)
    }

    private fun setActiveLocation(locationId: String) {
        subscription?.close()
        subscription = try {
            ditto.sync.registerSubscription(
                """
                    SELECT * FROM ${Order.COLLECTION_NAME}
                    WHERE _id.locationId = :locationId
                        AND createdAt > :TTL
                """.trimIndent(),
                mapOf("locationId" to locationId, "TTL" to startOfTodayIso())
            )
        } catch (error: Throwable) {
            reportSubscriptionFailure("subscribe orders", error)
        }
    }

    fun observeLocationOrders(locationId: String): Flow<List<Order>> =
        ditto.store.observeAsFlow(
            query = """
                SELECT * FROM ${Order.COLLECTION_NAME}
                WHERE _id.locationId = :locationId
                    AND createdAt > :TTL
            """.trimIndent(),
            args = mapOf("locationId" to locationId, "TTL" to startOfTodayIso())
        )

    suspend fun upsert(order: Order) {
        ditto.store.execute(
            """
                INSERT INTO ${Order.COLLECTION_NAME}
                DOCUMENTS (deserialize_json(:json))
                ON ID CONFLICT DO UPDATE_LOCAL_DIFF
            """.trimIndent(),
            mapOf("json" to order.dittoJsonString())
        )
    }

    suspend fun clearCart(order: Order) {
        if (order.cart.isEmpty()) return
        // UNSET target paths can't be parameterized in DQL, so the cart keys
        // (app-generated line-item UUIDs) are interpolated; all values use :named args.
        val unsetList = order.cart.keys.joinToString(", ") { "cart.\"$it\"" }
        ditto.store.execute(
            """
                UPDATE ${Order.COLLECTION_NAME}
                UNSET $unsetList
                WHERE _id.id = :id AND _id.locationId = :locationId
            """.trimIndent(),
            mapOf("id" to order.documentId.id, "locationId" to order.documentId.locationId)
        )
    }

    suspend fun reset(order: Order) {
        val createdAtNow = Clock.System.now().toDittoIsoString()
        val baseArgs = mapOf<String, Any>(
            "id" to order.documentId.id,
            "locationId" to order.documentId.locationId,
            "createdAt" to createdAtNow
        )
        val query = if (order.cart.isEmpty()) {
            """
                UPDATE ${Order.COLLECTION_NAME}
                SET createdAt = :createdAt
                WHERE _id.id = :id AND _id.locationId = :locationId
            """.trimIndent()
        } else {
            val unsetList = order.cart.keys.joinToString(", ") { "cart.\"$it\"" }
            """
                UPDATE ${Order.COLLECTION_NAME}
                SET createdAt = :createdAt
                UNSET $unsetList
                WHERE _id.id = :id AND _id.locationId = :locationId
            """.trimIndent()
        }
        ditto.store.execute(query, baseArgs)
    }

    suspend fun runEvictionIfDue() {
        val prefs = context.getSharedPreferences(EVICTION_PREFS, Context.MODE_PRIVATE)
        val now = Clock.System.now().toEpochMilliseconds()
        val last = prefs.getLong(LAST_EVICTION_KEY, 0L)
        if (now - last < TWENTY_FOUR_HOURS_MILLIS) return

        val ttl = startOfTodayIso()
        try {
            ditto.store.execute(
                "EVICT FROM ${Order.COLLECTION_NAME} WHERE createdAt <= :TTL",
                mapOf("TTL" to ttl)
            )
            prefs.edit().putLong(LAST_EVICTION_KEY, now).apply()
            Log.i("Eviction", "evicted orders with createdAt <= $ttl")
        } catch (error: Throwable) {
            Log.w("Eviction", error.message.orEmpty())
        }
    }

    companion object {
        private const val TWENTY_FOUR_HOURS_MILLIS = 24L * 60 * 60 * 1000
        private const val EVICTION_PREFS = "ditto_pos_eviction"
        private const val LAST_EVICTION_KEY = "v2.lastEvictionAt"
    }
}

private fun startOfTodayIso(): String {
    val tz = TimeZone.currentSystemDefault()
    return Clock.System.now().toLocalDateTime(tz).date.atStartOfDayIn(tz).toDittoIsoString()
}
